### PR TITLE
sql: support external row data spans for tables

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit teaches the fetcher machinery to read from the "external row data" which will allow us to support use cases like reading from a stand-by cluster in PCR or from backups. This is achieved by injecting the re-mapping step for all Scans, ReverseScans, and Gets before they get issued against the KV layer. At the moment, we will only support cases where the schemas of the two tables matches exactly, so we only need to remap the tenant and TableID prefix for each relevant key.

There is an open question of what the value of `tableoid` system column should be for this external row data. If it should match the ID of the external data, then we'd only need to remap the tenant prefix.

This is a polished version of https://github.com/dt/cockroach/commit/9af6eccf863c1881d6fbbd51520899fdfd22e33b from dt.

Epic: CRDB-37520

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
@dt a couple of questions:
- what should be the value of `tableoid` when reading from the external row data? If it's not important, then we could just get away with remapping the tenant prefix and adjust the IndexFetchSpec to have the "external" table ID so that keys would be almost constructed as we need.
- I'm currently hard-coding that IndexID will be the same between the two tables, is this a reasonable assumption?

**Comment by DrewKimball** [comment]
What's the reasoning for making this an execution-level change instead of (for example) remapping table references during planning time? Is it because we need to be able to read the original table at a different timestamp from the rest of the transaction?

**Review by yuzefovich — COMMENTED** [review]
This is probably a question for @dt, but I think that different timestamp might be part of it. Another possible justification for doing it this way is that we have easy control over which requests get issued against the external spans, so we can ensure that only a subset of read-only requests get sent. (Perhaps this could be done in other ways too - I'm just guessing about this.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-:-O1TCaO87EQWifndnL3O:b-85tnrs)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @dt)

**Comment by dt** [comment]
> What's the reasoning for making this an execution-level change instead of (for example) remapping table references during planning time?

Exactly what @yuzefovich said:

The fact that we need to fix the timestamps to a specific value, independent of the txn, with a different KV read is a big one I think, and we want to to be very, very, very sure we only ever do this peering into another tenant with specific RPCs that are known to be safe and read-only. Pointing the whole SQL engine at a table it doesn't own makes me very nervous but this narrowly scoped special case is pretty easy to reason about.

**Comment by DrewKimball** [comment]
>The fact that we need to fix the timestamps to a specific value, independent of the txn, with a different KV read is a big one I think, and we want to to be very, very, very sure we only ever do this peering into another tenant with specific RPCs that are known to be safe and read-only. Pointing the whole SQL engine at a table it doesn't own makes me very nervous but this narrowly scoped special case is pretty easy to reason about.

I don't know if I'm convinced by this argument. We definitely need to plumb a custom `AsOf` clause the way this PR does, but I'm suspicious of this low-level remapping when it seems like we could avoid it by just pointing at the actual table. There's already precedent for checking during planning whether a data source is being used in a valid context (e.g. mutation, locking, DDL etc.). If anything, I think this will be easier and less error-prone than doing it at a lower level.

**Comment by mgartner** [comment]
I have the same question as Drew.

One alternative is to present a new implementation of `catalog.TableDescriptor` that returns the translated table IDs in `GetID()` here (we might want to define a new interface with just the `GetID` method so the entire `catalog.TableDescriptor` interface does not have to be implemented): https://github.com/cockroachdb/cockroach/blob/66181b4fa59fcde749045d55718e7ee4115cd932/pkg/sql/span/span_builder.go#L48-L56

**Comment by yuzefovich** [comment]
There is an extensive discussion between Drew and David [here](https://cockroachlabs.slack.com/archives/C03JCUUSCD6/p1720733765648489?thread_ts=1720554297.766579&cid=C03JCUUSCD6) with Drew getting mostly on-board with the current IIUC.

**Comment by dt** [comment]
> translated table IDs in GetID()

To be clear, it isn't just a different _table_ ID: we need to use a whole different SQL _codec_ which is something that is very, very weird and unusual, and would generally be a bug if most code did it.

To drew's question: We could translate these spans earlier, but I hesitate to do that for a couple reasons: 

1) it only makes this one type -- data in another span of this cluster -- of externalized row data a little easier; if we had another type of externalized row data that needs to be fetched out of a backup on s3, or from a different _cluster_ over something other than kv scan reqs, it doesn't help us at all.

2) we don't _just_ need to change the span, but rather need to use a different request, specifically a fixed timestamp request. Right now the same block of code does both of these things: changes the span and uses a request that obeys the timestamp restriction. If we moved the span changing so that the default reader could just read from the normal spans field and it would appear to work, even if it forgot to apply the fixed ts reads, we could have a serious correctness issue. By tying the span redirection to the special, fixed timestamp requests, I think it'd be more obvious if we miss a case since it'll be reading completely the wrong (and empty) span, rather than what looks 99% right except when it happens to be executing on the wrong ts.

If you just don't like the "rekey" approach, I could see adding a second, parallel "external spans" field that we populate with a span builder using a different codec, but I'd prefer that we don't mix spans that are special and need to be read specially with the normal ones in the same field.

File: pkg/sql/rowenc/index_fetch.go
Comment (navsetlur) [inline pkg/sql/rowenc/index_fetch.go:61]: Could I get more insight into what MakeSQLCodec does? I understand generally it provides the details for the external table but how?

File: pkg/sql/row/kv_batch_fetcher.go
Comment (navsetlur) [inline pkg/sql/row/kv_batch_fetcher.go:338]: is bufalloc more efficient than make?

File: pkg/sql/row/kv_batch_fetcher.go
Comment (navsetlur) [inline pkg/sql/row/kv_batch_fetcher.go:341]: NIT: You could also simplify the end here with:
````
if cap(k) >= len(ext.NewPrefix)+len(suffix) {
    newK := k
else {
    newK := make([]byte, len(ext.NewPrefix)+len(suffix))
}

copy(newK[len(ext.NewPrefix):], suffix)
copy(newK, ext.NewPrefix)
````

File: pkg/sql/row/kv_batch_fetcher.go
Comment (navsetlur) [inline pkg/sql/row/kv_batch_fetcher.go:328]: NIT: You could also simplify the end here with:
````
if cap(k) >= len(ext.NewPrefix)+len(suffix) {
    newK := k
else {
    newK := make([]byte, len(ext.NewPrefix)+len(suffix))
}

copy(newK[len(ext.NewPrefix):], suffix)
copy(newK, ext.NewPrefix)
````

**Review by navsetlur — APPROVED** [review]
Looks good to me. It looks like the deeper design concerns have been addressed, and from a code standpoint I see no issues.

**Review by yuzefovich — COMMENTED** [review]
TFTR!

I'll wait for an approval from Drew or Marcus before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-:-O2AewWz-wWgpE7X4hgO:bkqr7x6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @dt, and @navsetlur)

___
*[`pkg/sql/row/kv_batch_fetcher.go` line 328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126875#gh-1683822146:-O2AbwApAyvBUUMzHKl9:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5d9bc89c2ed52990f7cc8d4c5b96c0697e97a41/pkg/sql/row/kv_batch_fetcher.go#L328)):*
<details><summary><i>Previously, navsetlur wrote…</i></summary><blockquote>

NIT: You could also simplify the end here with:
````
if cap(k) >= len(ext.NewPrefix)+len(suffix) {
    newK := k
else {
    newK := make([]byte, len(ext.NewPrefix)+len(suffix))
}

copy(newK[len(ext.NewPrefix):], suffix)
copy(newK, ext.NewPrefix)
````


</blockquote></details>

Good point, done.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 338 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126875#gh-1683816412:-O2Ac-AuEhdSRHZkg8iG:b-e2fe20) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5d9bc89c2ed52990f7cc8d4c5b96c0697e97a41/pkg/sql/row/kv_batch_fetcher.go#L338)):*
<details><summary><i>Previously, navsetlur wrote…</i></summary><blockquote>

is bufalloc more efficient than make?
</blockquote></details>

`bufalloc.ByteAllocator` is a utility struct that batches allocations of `[]byte` slices by allocating huge `[]byte` and then giving out sub-slices. The reasoning is that every allocation via `make` results in a separate allocation and a separate object that needs to be tracked by GC, but if we create one large slice and then divide it into, say, 100 smaller slices, we reduce both the overhead of making an allocation (allocating 100KiB slice is cheaper than allocating a hundred 1KiB slices) and of having to track 1 object instead of 100 for GC.
___
*[`pkg/sql/rowenc/index_fetch.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126875#gh-1683755019:-O2AcytDAW2brP3mS1Kw:b7sl5ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5d9bc89c2ed52990f7cc8d4c5b96c0697e97a41/pkg/sql/rowenc/index_fetch.go#L61)):*
<details><summary><i>Previously, navsetlur wrote…</i></summary><blockquote>

Could I get more insight into what MakeSQLCodec returns? I understand generally it provides the details for the external table but how? 
</blockquote></details>

First, some background (detailed information can be found at https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/encoding.md): we encode all table keys like `/<TableID>/<IndexID/<PK col1>/<PK col2/...`. Since we introduced multi-tenancy as a concept after having developed the DB for a few years, we needed to adjust this scheme to ensure separation of key spaces of different tenants, so we prepended the TenantID to all the keys, making overall encoding look like `/<TenantID>/<TableID>/<IndexID/<PK col1>/<PK col2/...` and to make it backwards compatible the system tenant's TenantID is encoded as nothing (i.e. empty).

`SQLCodec` is utility struct that prepends necessary `TenantID` prefix for all keys processed within a tenant. It was introduced throughout the whole codebase to ensure that we never try to construct a key without the SQLCodec so that we don't forget to prepend the TenantID.

To answer your question: in short, the job of the `SQLCodec` is to simply prepend the encoded `TenantID` for the tenant which this codec corresponds to.

**Review by DrewKimball — APPROVED** [review]
Code changes :lgtm:

> Reviewed 16 of 19 files at r1, 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-:-O2BpfBb0NSf8bd4nYto:be6b4rb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @navsetlur, and @yuzefovich)

___
*[`pkg/sql/row/external_row_data_test.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-O2Bp9D95e-rgf1vG_h_:-O2Bp9D95e-rgf1vG_ha:b-bag3kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cd9b97e095dc9c78a1c97e350f83c71550053ae/pkg/sql/row/external_row_data_test.go#L99)):*
> ```Go
> 				expected: [][]string{{"2"}},
> 			},
> 			{ // lookup join which might be served via the Streamer
> ```

Should we also add an index join case?

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-:-O2CTWYz9KkybFnAow9x:b-tlx7ko)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @dt, and @navsetlur)

___
*[`pkg/sql/row/external_row_data_test.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126875#-O2Bp9D95e-rgf1vG_h_:-O2CTSf2315qipwOSotk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cd9b97e095dc9c78a1c97e350f83c71550053ae/pkg/sql/row/external_row_data_test.go#L99)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we also add an index join case?
</blockquote></details>

Done.

**Comment by mgartner** [comment]
> To be clear, it isn't just a different table ID: we need to use a whole different SQL codec which is something that is very, very weird and unusual, and would generally be a bug if most code did it.

The span builder is already aware of codecs, so utilizing a "codec override" fetched from the descriptor should be straight forward.

> To drew's question: We could translate these spans earlier, but I hesitate to do that for a couple reasons:
>
>    1. it only makes this one type -- data in another span of this cluster -- of externalized row data a little easier; if we had another type of externalized row data that needs to be fetched out of a backup on s3, or from a different cluster over something other than kv scan reqs, it doesn't help us at all.
>
>    2. we don't just need to change the span, but rather need to use a different request, specifically a fixed timestamp request. Right now the same block of code does both of these things: changes the span and uses a request that obeys the timestamp restriction. If we moved the span changing so that the default reader could just read from the normal spans field and it would appear to work, even if it forgot to apply the fixed ts reads, we could have a serious correctness issue. By tying the span redirection to the special, fixed timestamp requests, I think it'd be more obvious if we miss a case since it'll be reading completely the wrong (and empty) span, rather than what looks 99% right except when it happens to be executing on the wrong ts.
>
> If you just don't like the "rekey" approach, I could see adding a second, parallel "external spans" field that we populate with a span builder using a different codec, but I'd prefer that we don't mix spans that are special and need to be read specially with the normal ones in the same field.

Thanks for the additional context! I'm still not fully convinced that the span builder is **not** the place to do this. It is solely responsible for translating logical spans into physical spans, so it seems like a more natural than coupling execution code with span translation.

I'm also confused how translating as implemented by this PR could work with DistSQL, which needs to know the physical spans in order to push execution operators to the correct nodes. 

Can you explain more about (1)? It's not obvious to me how moving the translation step would prevent these cases.

**Comment by yuzefovich** [comment]
> I'm also confused how translating as implemented by this PR could work with DistSQL, which needs to know the physical spans in order to push execution operators to the correct nodes.

That's a good point that I didn't really consider.

Depending on the deployment (either UA or external process / serverless) the behavior will be different. We need to know the "final" physical keys _and_ the dist sender range cache needs to be aware of those "external" ranges for us to make good physical planning decisions, and I believe with the current approach neither of the two will be in place, so DistSQL decisions will be effectively random. Making the key translation earlier, before the DistSQL physical planning, will address the first item, but there is still a question of somehow populating the DistSender range cache with information about the external ranges (if this external row data is within the same tenant, then it should be easy, otherwise we'd need to break the tenant boundary). Perhaps this is not a big issue for now - maybe we'll just disable DistSQL altogether for reads from the stand-by (?).

**Comment by mgartner** [comment]
@dt @msbutler @DrewKimball and I met today to discuss this a bit more. Some key points we discussed:

1. The urgency of supporting DistSQL in the read tenant of a PCR cluster is not clear, but it is likely that this read tenant will be used for OLAP-like queries which would benefit significantly from DistSQL.
2. In order to enable DistSQL on the read tenant, we'll need to make a similar translation between keyspaces when we partition spans. We discussed a high-level implementation plan where we build the translated spans earlier on in the execution process (similar to what is suggested above) and use those keys only in partitioning spans and when reading from tables. This will ensure that we aren't incorrectly translating spans across key spans when we don't want to (e.g., like writes) while reducing duplicate work in the translation.
3. There was initially some concern about getting table statistics to work correctly. There are two options:
  a. Reuse the replicated stats that are already in the other tenant. This would save us from having to recompute stats in the reader tenant.
  b. Run `CREATE STATISTICS` periodically as data is replicated. The table reader in the stats job should already be making the correct keyspace tanslation with the work merged in this PR.

**Pull request merged**
